### PR TITLE
Sort

### PR DESCRIPTION
**Pull request by mattwchun**
**Description of the problem this pull request fixes**
This feature allows users to sort based on rows in the DataTable. This will be done with vega sort data transform code added to the datatable spec.

Changes proposed in this pull request:
## \- sorting feature to allows users to sort based on rows in the DataTable.

*\* If submitting code for review: **

**This PR includes:**
- [ ] Tests
- [ ] functional comments
- [ ] high level description of any new components

**Steps to functionally test this:**
## 
## 
##

### CODE REVIEW

**Comment by AnjirHossain** [comment]
Just went through the QA testing flow, the pr didn't introduce any regressions and all the expected sorting functionality is present. 1+ functional!

File: src/assets/increasingSort.svg [hunk 1]
```diff
@@ -0,0 +1,4 @@
...
+<!--<text x="0" y="115" fill="#000000" font-size="5px" font-weight="bold" font-family="'Helvetica Neue', Helvetica, Arial-Unicode, Arial, Sans-serif">Created by Chris</text><text x="0" y="120" fill="#000000" font-size="5px" font-weight="bold" font-family="'Helvetica Neue', Helvetica, Arial-Unicode, Arial, Sans-serif">from the Noun Project</text>-->
+<!-- changed the height in viewBox to 125 to 100 -->
+</svg>
```
Comment (arvind) [inline src/assets/increasingSort.svg]: These two icons are pretty difficult to identify in the `HoverField`. I've had more luck with the [FontAwesome icon set](http://fontawesome.io/icons/). You can find SVGs for them [here](https://github.com/encharm/Font-Awesome-SVG-PNG/tree/master/black/svg). 

With regards to _which_ icon to show (and I'm now referring to the corresponding FontAwesome icons here), let's start by showing `sort` (double arrow) to indicate that one _can_ sort a field, but that it has _not yet_ been sorted. Once a user clicks the icon, the icon should then switch to indicate the current sort order (i.e., should show ascending if ascending and descending if descending). Let's also use the icon most appropriate for the field type (i.e., `sort-alpha-` for categorical fields, and `sort-numeric-` for the other two).

File: src/js/components/pipelines/HoverField.jsx [hunk 2]
```diff
@@ -110,15 +117,41 @@ var HoverField = React.createClass({
...
+          <Icon onClick={this.sortValues}
+            glyph={assets['increasingSort']} width="10" height="10" /> ) : (
+          <Icon onClick={this.sortValues}
+            glyph={assets['decreasingSort']} width="10" height="10" /> );
 
     field = field ? (
       <div>
         <FieldType field={field} />
         {field.name}
+        {incOrDecIcon}
```
Comment (arvind) [inline src/js/components/pipelines/HoverField.jsx]: Can you extract all the changes you've made to this component into a separate `SortField.jsx` component? The goal would be to create a fairly standalone/self-contained component, much like we did with `FieldType.jsx`. Unless I'm mistaken, the `SortField` component should not require any local state (i.e., `this.state` within the react component itself). Instead, we should be able to drive it entirely based on what is stored in the `dataset._sort` in the store.

File: src/js/ctrl/export.js [hunk 3]
```diff
@@ -55,7 +56,12 @@ exporter.dataset = function(state, internal, id) {
       json2csv({data: values, del: format === 'tsv' ? '\t' : ','}) : values;
   }
 
-  return spec;
+  if(sort == undefined) {
+    return spec;
+  } else {
+    spec['transform'] = sort;
```
Comment (arvind) [inline src/js/ctrl/export.js]: `spec.transform` will suffice here. Stylistically, we prefer to only use square bracket notation when the key is an invalid identifier (e.g., contains symbols).

File: src/js/ctrl/export.js [hunk 4]
```diff
@@ -230,6 +236,29 @@ exporter.axe = exporter.legend = function(state, internal, id) {
...
 /**
+  * Method that builds the vega sort data transform code from
+  * the current dataset.
+  *
+  * @param  {object} dataset The current dataset
+  * @returns {object} undefined if _sort not in dataset and the
+  * vega data transform code to be appended to the vega spec to * the dataset
+  */
+function getSort(dataset) {
+  if (!('_sort' in dataset)) {
```
Comment (arvind) [inline src/js/ctrl/export.js]: Simply `if (!dataset._sort)` should be fine here.

File: src/js/components/pipelines/HoverField.jsx [hunk 5]
```diff
@@ -110,15 +117,41 @@ var HoverField = React.createClass({
...
+
+  if (valuesInc == null) {
+    // first click default: increasing
+    valuesInc = true;
+  } else {
+    valuesInc = !valuesInc;
+  }
+  this.setState({ valuesInc : valuesInc });
+
+  incOrDec = valuesInc ? 'inc' : 'dec';
```
Comment (arvind) [inline src/js/components/pipelines/HoverField.jsx]: Let's use `asc` or `desc` instead to conform to typically expected sort order values. (Be sure to update these throughout).

File: src/js/components/pipelines/HoverField.jsx [hunk 6]
```diff
@@ -110,15 +117,41 @@ var HoverField = React.createClass({
...
+      id = props.dsId,
+      incOrDec = null;
+
+  if (valuesInc == null) {
+    // first click default: increasing
+    valuesInc = true;
+  } else {
+    valuesInc = !valuesInc;
+  }
+  this.setState({ valuesInc : valuesInc });
```
Comment (arvind) [inline src/js/components/pipelines/HoverField.jsx]: Be sure to run `npm run lint` to catch errors in code style like the additional spaces here.

**Comment by arvind** [comment]
Hi @mattwchun, this looks good for the most part. I've left you a handful of comments throughout, and will merge this once they're handled. Thanks!

**Comment by arvind** [comment]
Hi @mattwchun, great work overall! I would take a look at bf80429 to see the ways in which I iterated on and refined your work. In particular, I extracted any hardcoded strings that were reused in multiple places into a standalone `constants` file. 

I also simplified the code path in the `SortField` component. This was motivated partly by an error-case I observed. Previously, if no fields were sorted and I clicked to sort `fieldA` it would be sorted in ascending order. Then, if I clicked to sort `fieldB`, that would flip to descending order on the first click. This is because you were considering the sort state to flip the order, rather than looking at if the fields matched. 

Finally, some stylistic changes -- e.g., preferring a more concise store shape (`._sort.field` and `._sort.order` rather than `._sort.sortField` and `._sort.sortOrder`). 

Hope that helps guide your future PRs! Let me know if you have any questions. Great work overall!

**Pull request merged**